README.md
------------------------------------------------------------------------
yPcalc - An IPv4 and IPv6 subnet calculator
========================================================================
------------------------------------------------------------------------

1.  Description
2.  Requirements
3.  Installation
4.  Usage
5.  License
6.  Support
7.  Authors

1. Description
--------------

yPcalc is a simple Python based IPv4 and IPv6 calculator. It's output
is easy to parse which makes it possible to intergrate it to shell
scripts.

2. Requirements
---------------

- Python 2.7 (<http://python.org>);
- Python `ipcalc` module (<https://pypi.python.org/pypi/ipcalc>).

3. Installation
---------------

- Download and install [ipcalc](https://pypi.python.org/pypi/ipcalc/)
Python module. I recommend to use the latest one from the
[GitHub](https://github.com/tehmaze/ipcalc/archive/master.zip) and
install it by running the setup script in command prompt:

        $ python setup.py install

- Download the latest release from
<https://github.com/gynter/yPcalc/releases/latest>
and extract the archive to the chosen destination directory.

    Development files can be browsed via web browser or can be optained
    from a git repository <https://github.com/gynter/yPcalc>.

    *Note: You can also get the sources using the
    [GIT](http://git-scm.com/book/en/Getting-Started-Installing-Git) and
    cloning the `https://github.com/gynter/yPcalc.git` repository.*

- Install it by running the the setup script in command prompt:

        $ python setup.py install

*Note: You must have administrator/super user privileges
for the installation.*

4. Usage
--------

yPcalc is a command line software, therefore all actions are performed
in command prompt.

    $ ypcalc -h
    usage: ypcalc [-h] [-V] [-L] {show,check} ...

    yPcalc - An IPv4 and IPv6 subnet calculator.

    optional arguments:
      -h, --help     show this help message and exit
      -V, --version  show program's version number and exit
      -L, --license  show program's license end exit

    commands:
      {show,check}   see {command} -h for more help
        show         show the IP address information
        check        check if IP address exists in the network. Exits with code 1
                     if not found

It has two commands:

- `show` displays the information about the IP or network address;
- `check` performs a lookup to see if the IP address exists in the network.

See `ypcalc {command} -h` for command specific arguments.

IPv4 examples:
   
    $ ypcalc show 192.0.3.171/27
    Address:     192.0.3.171/27
    Netmask:     255.255.255.224 = 27
    Network:     192.0.3.160/27
    HostMin:     192.0.3.161
    HostMax:     192.0.3.190
    Broadcast:   192.0.3.191
    Hosts/Net:   32 UNKNOWN
    IPv6 repr:   2002:c000:03ab:0000:0000:0000:0000:0000
    PTR RR name: 171.3.0.192.in-addr.arpa
    IP version:  4

    $ ypcalc check 192.0.3.164 192.0.3.171/27
    ypcalc: info: IP 192.0.3.164 exists in network 192.0.3.171/27

    $ ypcalc check 192.0.3.192 192.0.3.171/27
    ypcalc: error: IP 192.0.3.192 does not exist in network 192.0.3.171/27

IPv6 examples:

    $ ypcalc show 2002:c000:022a::/29
    Address:     2002:c000:022a:0000:0000:0000:0000:0000/29
    Netmask:     ffff:fff8:0000:0000:0000:0000:0000:0000 = 29
    Network:     2002:c000:0000:0000:0000:0000:0000:0000/29
    HostMin:     2002:c000:0000:0000:0000:0000:0000:0001
    HostMax:     2002:c007:ffff:ffff:ffff:ffff:ffff:fffe
    Hosts/Net:   633825300114114700748351602688 UNKNOWN
    IPv4 repr:   192.0.2.42
    PTR RR name: 0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.a.2.2.0.0.0.0.c.2.0.0.2.ip6.arpa
    IP version:  6

    $ ypcalc check 2002:c006:022a:: 2002:c000:022a::/29
    ypcalc: info: IP 2002:c006:022a:: exists in network 2002:c000:022a::/29

    $ ypcalc check 2002:c009:022a:: 2002:c000:022a::/29
    ypcalc: error: IP 2002:c009:022a:: does not exist in network 2002:c000:022a::/29

When using `output filter` then only specific values are displayed,
without the caption prefix:

    $ ypcalc show --reverse --netmask 2002:c000:022a::/29
    ffff:fff8:0000:0000:0000:0000:0000:0000
    0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.a.2.2.0.0.0.0.c.2.0.0.2.ip6.arpa

The values are ordered as shown in `ypcalc show -h` regardless of the
sequence of the arguments.

5. License
----------

This software is licensed as described in the file `LICENSE.md`, which
You should have received as part of this distribution. The terms are
also available at
<https://github.com/gynter/yPcalc/blob/master/LICENSE.md>.

6. Support
----------

Issue tracker can be found at
<https://github.com/gynter/yPcalc/issues>.

7. Authors
----------

- Günter Kits (gynter@kits.ee)


CHANGELOG.md
------------------------------------------------------------------------
CHANGELOG yPcalc
========================================================================
------------------------------------------------------------------------

v1.0.0 (2015-07-16)
-------------------

- First release, bump version.

v0.1.0-beta.5 (2015-07-16)
--------------------------

- Fix README;
- Fix copyright dates.

v0.1.0-beta.4 (2014-10-16)
--------------------------

- Pushed year to 2014 in LICENSE.md and headers;
- Upgraded Distribute installation script (ez_setup);
- Added Network to IPv6 output;
- Updated IPv6 example.

v0.1.0-beta.3 (2013-12-23)
--------------------------

- Mistakenly added wrong classifiers to setup.py.

v0.1.0-beta.2 (2013-12-23)
--------------------------

- Fixed versioning;
- Small changes in .gitignore;
- Added classifiers to setup.py and fixed indentation;
- Added Bootstrap setuptools installation.

v0.1.0-beta.1 (2013-10-31)
--------------------------

- Using Semantic Versioning system for versioning;
- Changes in documentation.

v0.1.0-beta (2013-09-19)
------------------------

- Start of yPcalc project.


src/ypcalc/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
__LICENSE__ = """
yPcalc - An IPv4 and IPv6 subnet calculator.

Copyright (c) 2013 Günter Kits

Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal in the Software without restriction, 
including without limitation the rights to use, copy, modify, merge, 
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

The above copyright notice and this permission notice shall be 
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS 
BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN 
ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
SOFTWARE.
"""

import sys
from os.path import basename
from copy import copy
import argparse

__PROGRAM__ = "yPcalc"
__VERSION__ = "1.0.0"
__DESCRIPTION__ = "An IPv4 and IPv6 subnet calculator."
__EXECUTABLE__ = basename(sys.argv[0])


def exit(msg='', code=0):
    if msg:
        if code:
            print >>sys.stderr, "%s: error: %s" % (__EXECUTABLE__, msg)
        else:
            print "%s: info: %s" % (__EXECUTABLE__, msg)
    sys.exit(code)


try:
    from ipcalc import IP, Network
except ImportError:
    exit("failed to load ipcalc module, please see the documentation", 1)


class LicenseAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        print __LICENSE__
        exit(code=1)


def split_len(seq, length):
    return [seq[i:i+length] for i in range(0, len(seq), length)]

def get_network(network_addr):
    if not network_addr:
        print exit("no network IP address defined", 1)
    try:
        return Network(network_addr)
    except ValueError as e:
        print exit(e, 1)

def repr_addr(addr, args):
    if args.bin:
        addr = addr.bin()
        if len(addr) == 32 and not args.nosep:
            addr = '.'.join(split_len(addr, 8))
        elif len(addr) == 128 and not args.nosep:
            addr = '.'.join(split_len(addr, 16))
    elif args.hex:
        addr = addr.hex()
        if len(addr) == 8 and not args.nosep:
            addr = ':'.join(split_len(addr, 2))
        elif len(addr) == 32 and not args.nosep:
            addr = ':'.join(split_len(addr, 4))
    return addr

def _(caption, value, args, short=False, addr=False):
    raddr = ""
    if addr:
        if isinstance(addr, bool):
            raddr = repr_addr(value, args)
        else:
            raddr = repr_addr(addr, args)
    if short:
        return raddr if addr else value
    if caption:
        caption = ':'.join([caption, ''])
        if args.bin or args.hex:
            return "%s%s%s" % (caption.ljust(13), str(value).ljust(50),
                               raddr)
        return "%s%s" % (caption.ljust(13), str(value))

def command_show(args):
    fields = ['address', 'netmask', 'network', 'min', 'max', 'broadcast',
              'hosts', 'info', 'ipv4', 'ipv6', 'reverse', 'ipversion']

    n = get_network(args.netaddr)
    short = any(args.__dict__.get(x, False) for x in fields)

    filters = [
        ('Address', n, IP(str(n)), 0),
        ('Netmask', "%s = %s" % (n.netmask(), n.mask), n.netmask(), 0),
        ('Network', "%s/%s" % (n.network(), n.mask), n.network(), 0),
        ('HostMin', n.host_first(), True, 0),
        ('HostMax', n.host_last(), True, 0),
        ('Broadcast', n.broadcast(), True, 4),
        ('Hosts/Net', n.size() if short else "%s %s" % (n.size(), n.info()), False, 0),
        ('', n.info(), False, 0),
        ('IPv4 repr', n.to_ipv4(), True, 6),
        ('IPv6 repr', n.to_ipv6(args.type), True, 4),
        ('PTR RR name', n.to_reverse(), False, 0),
        ('IP version', n.version(), False, 0)
    ]

    for idx, f in enumerate(fields):
        d = filters[idx]
        o = _(d[0], d[1], args, short, d[2])
        if o and ((n.version() == d[3]) or not d[3]) and (args.__dict__.get(f, False) and short or not short):
            print o

def command_check(args):
    network = get_network(args.netaddr)
    # ipcalc module seems to have a bug/feature somewhere, so we need
    # to get the real network, where the netaddr resides.
    network = get_network("%s/%s" % (network.network(), network.mask))
    try:
        if network.has_key(args.addr):
            exit("IP %s exists in network %s" % (args.addr, args.netaddr))
    except ValueError as e:
        print exit(e, 1)
    exit("IP %s does not exist in network %s" % (args.addr, args.netaddr), 1)


def run():
    parser_desc = "%s - %s" % (__PROGRAM__, __DESCRIPTION__)

    # Main arguments.
    p_ap = argparse.ArgumentParser(add_help=False)
    p_ap.add_argument('-V', '--version', action='version',
                      version='%s %s' % (__PROGRAM__, __VERSION__,))
    p_ap.add_argument('-L', '--license', action=LicenseAction, nargs=0,
                      help="show program's license end exit")

    # Main parser.
    p = argparse.ArgumentParser(description=parser_desc, parents=[p_ap])

    # Common arguments for commands.
    c_ap = argparse.ArgumentParser(add_help=False)
    c_ap_g = c_ap.add_argument_group(title='output representation',
                                     description='display the IP \
addresses in different format')
    c_ap_g.add_argument('--bin', action='store_true', default=False,
                        help='binary representation of the IP addresses')
    c_ap_g.add_argument('--hex', action='store_true', default=False,
                        help='hexadecimal representation of the IP \
addresses')
    c_ap_g.add_argument('--nosep', action='store_true', default=False,
                   help='do not add separator (dot for IPv4 and colon \
for IPv6) when displayng the address in different representation')

    # Sub-parsers/commands.
    sp = p.add_subparsers(title='commands', help='see {command} -h for \
more help')
    
    # Show command and arguments.
    desc = "show the IP address information"
    c = sp.add_parser('show', parents=[p_ap, c_ap], description=desc,
                      help=desc)
    a = c.add_argument_group(title='output filter',
                             description='show only certain short info, \
ordered as defined below')
    a.add_argument('--address', action='store_true', default=False,
                   help='show IP address')
    a.add_argument('--netmask', action='store_true', default=False,
                   help='show network mask address')
    a.add_argument('--network', action='store_true', default=False,
                   help='show network address (IPv4 only)')
    a.add_argument('--min', action='store_true', default=False,
                   help='show the first useable IP address of the network')
    a.add_argument('--max', action='store_true', default=False,
                   help='show the last useable IP address of the network')
    a.add_argument('--broadcast', action='store_true', default=False,
                   help='show broadcast address (IPv4 only)')
    a.add_argument('--hosts', action='store_true', default=False,
                   help='show the count of the hosts in network')
    a.add_argument('--info', action='store_true', default=False,
                   help='show IANA allocation information for the IP \
address')
    a.add_argument('--ipv4', action='store_true', default=False,
                   help='show the IPv4 IP address representation of an \
IPv6 address. Works for IPv4-compat, IPv4-mapped, and 6-to-4 addresses')
    a.add_argument('--ipv6', action='store_true', default=False,
                   help='show the IPv6 IP address representation an IPv4 \
address. See --type option.')
    a.add_argument('--reverse', action='store_true', default=False,
                   help='show the DNS RR name field of PTR record')
    c.add_argument('netaddr', metavar='netaddr', action='store',
                        help='an IP address or a network (with IP or CIDR \
netmask)')
    a.add_argument('--ipversion', action='store_true', default=False,
                   help='show the IP version')
    c.add_argument('--type', action='store', default='6-to-4',
                   choices=['6-to-4', 'compat', 'mapped'],
                   help='address type for IPv6 representation \
(default: 6-to-4)')
    c.set_defaults(func=command_show)

    # Check command and arguments.
    desc = "check if IP address exists in the network. Exits with code \
1 if not found"
    c = sp.add_parser('check', parents=[p_ap], description=desc,
                      help=desc)
    c.add_argument('addr', metavar='addr', action='store',
                   help='an IP address')
    c.add_argument('netaddr', metavar='netaddr', action='store', 
                   help='a network (with IP or CIDR netmask)')
    c.set_defaults(func=command_check)

    # Parse the arguments and call the command.
    args = p.parse_args()
    if args.func:
        args.func(args)

if __name__ == '__main__':
    run()
